src/shape/bar.js

```diff
@@ -177,7 +177,8 @@ extend(ChartInternal.prototype, {
 		const barW = $$.getBarW(axis, barTargetsNum);
 		const barX = $$.getShapeX(barW, barTargetsNum, barIndices, !!isSub);
 		const barY = $$.getShapeY(!!isSub);
-		const barOffset = $$.getShapeOffset($$.isBarType, barIndices, !!isSub);
+		const shapeOffsetData = $$.getShapeOffsetData($$.isBarType);
+		const barOffset = $$.getShapeOffset(shapeOffsetData, barIndices, !!isSub);
```

Comment:
`.getShapeOffsetData()` is called to provide data parameter consumed by `.getShapeOffset()` only.

IMO, will be more simplifying call `.getShapeOffsetData()` within `.getShapeOffset()` as follows.
```js
// - const shapeOffsetData = $$.getShapeOffsetData($$.isBarType);
const barOffset = $$.getShapeOffset($$.isBarType, barIndices, !!isSub);

getShapeOffset(typeFilter, indices, isSub) {
	const $$ = this;
	
	const {shapeOffsetTargets, indexMapByTargetId} = $$.getShapeOffsetData(typeFilter);
```

Reply:
The problem is that `getShapeOffset` gets called for every single data point. Doing these recalculations for every point is the slow part, and why calling this outside is such a significant speed improvement

Reply:
That's right that `getShapeOffset` is called for every point, but is calling the returned function.
So, moving the call of `.getShapeOffsetData()` within `getShapeOffset` is not making to call repeatedly.

```js
getShapeOffset(typeFilter, indices, isSub) {
	...
	const {shapeOffsetTargets, indexMapByTargetId} = $$.getShapeOffsetData(typeFilter);

	// the returned function is called for each data point
	return (d, idx) => { ... };
}
```
Profiling also shows no significant different.
(I did tests with `columns: [["data1", 30, 200, 100],["data2", 30, 00, 100]]) / type: "bar"`.

- The above graph is when `generateGetBarPoints` is called within `getShapeOffset`
- And the latter is the PR version, which `generateGetBarPoints` is called and passed as argument to `getShapeOffset`.
> each (anonymous) is the call of returned function from `getShapeOffset()`

![image](https://user-images.githubusercontent.com/2178435/67833539-46681280-fb28-11e9-8cf2-f7ccd90defb4.png)

Reply:
In my testing I found this wasn't the case for grouped charts - `getShapeOffset` itself could be called numerous extra times. Try adding `groups: [["data1", "data2"]]` and you should see what I mean (I'll try to get an example for you soon)

Reply:
The returned function `.getShapeOffset()`, is used from `.generateDrawLine()` and `.updateCircleY()` (in case of bar type, `.generateGetBarPoints()`).

```js
generateDrawLine(lineIndices, isSub) {
	const getPoints = $$.generateGetLinePoints(lineIndices, isSub);

	const yValue = (d, i) => ($$.isGrouped(d.id) ?
		getPoints(d, i)[0][1] :
		yScaleGetter.call($$, d.id)($$.getBaseValue(d))
	);

updateCircleY() {
	const getPoints = $$.generateGetLinePoints($$.getShapeIndices($$.isLineType), false);

	return $$.isGrouped(id) ?
		getPoints(d, i)[0][1] :
		$$.getYScale(id)($$.getBaseValue(d));
},
```

Even though as my proposed review comment, the call counts are same as the current PR with the use of `data.groups` option.

I'll be on the right before to bump new release, so I'll be merge the PR with some changes.
If you see some perf degradation, then plz send PR for that.